source/ranges.tex

```diff
@@ -511,49 +572,66 @@
 \pnum
 The name \tcode{ranges::rend} denotes a customization point
 object\iref{customization.point.object}.
-Given a subexpression \tcode{E} and
-an lvalue \tcode{t} that denotes the same object as \tcode{E},
-if \tcode{E} is an rvalue and
-\tcode{enable_safe_range<remove_cvref_t<decltype((E))>>} is \tcode{false},
-\tcode{ranges::rend(E)} is ill-formed.
-Otherwise,
-\tcode{ranges::rend(E)} is
-expression-equivalent to:
+
+\pnum
+Given a subexpression \tcode{E} with type \tcode{T},
+let \tcode{t} be an lvalue that
+denotes the same object as \tcode{E} if \tcode{E} is a glvalue, and
+otherwise denotes the result of applying
+the temporary materialization conversion\iref{conv.rval} to \tcode{E}.
+Then:
+
 \begin{itemize}
 \item
-  \tcode{\placeholdernc{decay-copy}(t.rend())}
-  if it is a valid expression and its type \tcode{S} models
-\begin{codeblock}
-sentinel_for<decltype(ranges::rbegin(E))>
-\end{codeblock}
+  If \tcode{E} is an rvalue and
+  \tcode{enable_safe_range<remove_cv_t<T>>} is \tcode{false},
+  \tcode{ranges::rend(E)} is ill-formed.
 
 \item
-  Otherwise, \tcode{\placeholdernc{decay-copy}(rend(t))} if it is a valid
-  expression and its type \tcode{S} models
-\begin{codeblock}
-sentinel_for<decltype(ranges::rbegin(E))>
-\end{codeblock}
-  with overload
-  resolution performed in a context that includes the declaration:
+  Otherwise, if \tcode{T} is an array type\iref{basic.compound} and
+  \tcode{remove_all_extents_t<T>} is an incomplete type,
+  \tcode{ranges::rend(E)} is ill-formed with no diagnostic required.
+
+\item
+  Otherwise, if \tcode{\placeholdernc{decay-copy}(t.rend())}
+  is a valid expression whose type models
+  \tcode{\libconcept{sentinel_for}<decl\-type(ranges::rbegin(E))>}
+  then \tcode{ranges::rend(E)} is expression-equivalent to
+  \tcode{\placeholdernc{decay-copy}(t.rend(\brk{}))}.
+
+\item
+  Otherwise, if \tcode{T} is a class or enumeration type and
+  \tcode{\placeholdernc{decay-copy}(rend(t))}
+  is a valid expression whose type models
+  \tcode{\libconcept{sentinel_for}<decltype(ranges::rbegin(E))>}
+  with overload resolution performed in a context in which
+  unqualified lookup for \tcode{rend} finds only the declarations
 \begin{codeblock}
-template<class T> void rend(T&&) = delete;
+void rend(auto&) = delete;
+void rend(const auto&) = delete;
 \end{codeblock}
-  and does not include a declaration of \tcode{ranges::rend}.
+  then \tcode{ranges::rbegin(E)} is expression-equivalent to
```

Comment:
Woops - this is copy pasta in the ingress paper; we aren't specifying `ranges::rbegin(E)` here, we're specifying `ranges::rend(E)`.
```suggestion
  then \tcode{ranges::rend(E)} is expression-equivalent to
```